api.go

```diff
@@ -0,0 +1,226 @@
+	}
+	if err := httpServer.ListenAndServe(); err != nil {
+		panic(err)
+	}
+}
+
+type Server struct {
+	client temporalclient.Client
+}
+
+type Review struct {
+	Time    string `json:"time"`
+	User    string `json:"user"`
+	Message string `json:"message"`
+	Status  string `json:"status"`
+}
+
+type Workflow struct {
+	StartTime string   `json:"startTime"`
+	ID        string   `json:"id"`
+	RunID     string   `json:"runId"`
+	User      string   `json:"user"`
+	Action    string   `json:"action"`
+	Reviews   []Review `json:"reviews"`
+	Status    string   `json:"status"`
+}
+
+type ListOpenWorkflowResponse struct {
+	Workflows []Workflow `json:"workflows"`
+}
+
+func (s *Server) ListOpenWorkflow(w http.ResponseWriter, r *http.Request) {
+	res, err := s.client.ListOpenWorkflow(r.Context(), &workflowservice.ListOpenWorkflowExecutionsRequest{Filters: &workflowservice.ListOpenWorkflowExecutionsRequest_TypeFilter{
+		TypeFilter: &filter.WorkflowTypeFilter{Name: workflow.WorkflowName},
+	}})
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	ret := &ListOpenWorkflowResponse{
+		Workflows: make([]Workflow, len(res.Executions)),
+	}
+	for idx, exec := range res.Executions {
+		resp, err := s.client.QueryWorkflow(r.Context(), exec.Execution.WorkflowId, exec.Execution.RunId, "getState")
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		var result workflow.State
+		if err := resp.Get(&result); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+
+		}
+
+		reviews := make([]Review, len(result.Reviews))
+		for ridx, r := range result.Reviews {
+			reviews[ridx] = Review{
+				Time:    r.Timestamp.String()[:19],
+				Message: r.Message,
+				User:    r.Author,
+				Status:  r.Status.String(),
+			}
+		}
+		status := "open"
+		if result.IsApproved() {
+			status = "approve"
+		} else if result.IsLocked() {
+			status = "locked"
+		}
+
+		ret.Workflows[idx] = Workflow{
+			StartTime: exec.StartTime.String()[:19],
+			ID:        exec.Execution.WorkflowId,
+			RunID:     exec.Execution.RunId,
+			User:      result.User,
+			Action:    result.Action,
+			Reviews:   reviews,
+			Status:    status,
+		}
+	}
+
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(ret)
+}
+
+type ExecuteWorkflowRequest struct {
+	Time   string `json:"time"`
+	User   string `json:"user"`
+	Action string `json:"action"`
+}
+
+func (s *Server) ExecuteWorkflow(w http.ResponseWriter, r *http.Request) {
+	in := &ExecuteWorkflowRequest{}
+	if err := json.NewDecoder(r.Body).Decode(in); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Initialize the state.
+	state := &workflow.State{
+		Action: in.Action,
+		User:   in.User,
+	}
+
+	opts := temporalclient.StartWorkflowOptions{TaskQueue: workflow.QueueName}
+	we, err := s.client.ExecuteWorkflow(r.Context(), opts, workflow.WorkflowName, state)
```

Comment:
Would you use reusable workflow ids here in a more complete version? Probably would want to ensure that no two workflows are doing the same thing.

Reply:
Enhancement: application should be aware that same request already exists and prevent duplicate creation.

Reply:
Can view all pending actions, whichever one approved first wins.